lib/jekyll/tags/post_url.rb

```diff
@@ -61,44 +61,86 @@ class PostUrl < Liquid::Tag
       def initialize(tag_name, post, tokens)
         super
         @orig_post = post.strip
-        begin
-          @post = PostComparer.new(@orig_post)
-        rescue StandardError => e
-          raise Jekyll::Errors::PostURLError, <<~MSG
-            Could not parse name of post "#{@orig_post}" in tag 'post_url'.
-             Make sure the post exists and the name is correct.
-             #{e.class}: #{e.message}
-          MSG
-        end
+
+        @template = Liquid::Template.parse(@orig_post) if use_template?
       end
 
       def render(context)
         @context = context
-        site = context.registers[:site]
+        site = @context.registers[:site]
+        orig_post_expanded = expand_orig_post
+
+        begin
+          post = PostComparer.new(orig_post_expanded)
+        rescue StandardError => e
+          raise Jekyll::Errors::PostURLError,
+                could_not_parse_error_message(orig_post_expanded, e)
+        end
 
         site.posts.docs.each do |document|
-          return relative_url(document) if @post == document
+          return relative_url(document) if post == document
         end
 
         # New matching method did not match, fall back to old method
         # with deprecation warning if this matches
 
         site.posts.docs.each do |document|
-          next unless @post.deprecated_equality document
+          next unless post.deprecated_equality document
 
-          Jekyll::Deprecator.deprecation_message(
-            "A call to '{% post_url #{@post.name} %}' did not match a post using the new " \
-            "matching method of checking name (path-date-slug) equality. Please make sure " \
-            "that you change this tag to match the post's name exactly."
-          )
+          Jekyll::Deprecator.deprecation_message(create_deprecation_message(post.name))
           return relative_url(document)
         end
 
-        raise Jekyll::Errors::PostURLError, <<~MSG
-          Could not find post "#{@orig_post}" in tag 'post_url'.
+        raise Jekyll::Errors::PostURLError, could_not_find_error_message(orig_post_expanded)
+      end
+
+      private
+
+      def use_template?
+        content = @orig_post
+        return false if content.nil? || content.empty?
+
+        content.include?("{%") || content.include?("{{")
+      end
+
+      def expand_orig_post
+        return @orig_post if @template.nil?
+
+        @template.render(@context)
+      end
+
+      # if there is liquid rendering besides a simple constant, adds
+      # "(from input \"#{@orig_post}\")" to let maintainer know where and why it failed
+      def determine_post_string(orig_post_expanded)
+        if orig_post_expanded == @orig_post
+          "\"#{orig_post_expanded}\""
+        else
+          "\"#{orig_post_expanded}\" (from input \"#{@orig_post}\")"
+        end
+      end
```

Comment:
Ruby focuses on readability (and aesthetics). The following is equivalent:

```suggestion
      # if there is liquid rendering besides a simple constant, adds
      # `(from input "#{@orig_post}")` giving the user some context.
      def determine_post_string(orig_post_expanded)
        if orig_post_expanded == @orig_post
          orig_post_expanded.inspect
        else
          "#{orig_post_expanded.inspect} (from input #{@orig_post.inspect})"
        end
      end
```